File: snippets/go/parsing_inputs.go
package example

import (
	"log"
	"strconv"

	"github.com/breez/breez-sdk-liquid-go/breez_sdk_liquid"
)

func ParseInput(sdk *breez_sdk_liquid.BindingLiquidSdk) {
	// ANCHOR: parse-inputs
	input := "an input to be parsed..."

	if input, err := sdk.Parse(input); err == nil {
		switch inputType := input.(type) {
		case breez_sdk_liquid.InputTypeBitcoinAddress:
			log.Printf("Input is Bitcoin address %s", inputType.Address.Address)

		case breez_sdk_liquid.InputTypeBolt11:
			amount := "unknown"
			if inputType.Invoice.AmountMsat != nil {
				amount = strconv.FormatUint(*inputType.Invoice.AmountMsat, 10)
			}
			log.Printf("Input is BOLT11 invoice for %s msats", amount)

		case breez_sdk_liquid.InputTypeLnUrlPay:
			log.Printf("Input is LNURL-Pay/Lightning address accepting min/max %d/%d msats",
				inputType.Data.MinSendable, inputType.Data.MaxSendable)

		case breez_sdk_liquid.InputTypeLnUrlWithdraw:
			log.Printf("Input is LNURL-Withdraw for min/max %d/%d msats",
				inputType.Data.MinWithdrawable, inputType.Data.MaxWithdrawable)

		default:
			// Other input types are available
		}
	}
	// ANCHOR_END: parse-inputs
}

func ConfigureParsers() (*breez_sdk_liquid.BindingLiquidSdk, error) {
	// ANCHOR: configure-external-parser
	mnemonic := "<mnemonic words>"

	// Create the default config, providing your Breez API key
	breezApiKey := "<your-Breez-API-key>"
	config, err := breez_sdk_liquid.DefaultConfig(breez_sdk_liquid.LiquidNetworkMainnet, &breezApiKey)
	if err != nil {
		return nil, err
	}

	// Configure external parsers
	parsers := []breez_sdk_liquid.ExternalInputParser{
		{
			ProviderId: "provider_a",
			InputRegex: "^provider_a",
			ParserUrl:  "https://parser-domain.com/parser?input=<input>",
		},
		{
			ProviderId: "provider_b",
			InputRegex: "^provider_b",
			ParserUrl:  "https://parser-domain.com/parser?input=<input>",
		},
	}
	config.ExternalInputParsers = &parsers

	connectRequest := breez_sdk_liquid.ConnectRequest{
		Config:   config,
		Mnemonic: mnemonic,
	}

	sdk, err := breez_sdk_liquid.Connect(connectRequest)

	return sdk, err
	// ANCHOR_END: configure-external-parser
}

File: snippets/rust/src/parsing_inputs.rs
use anyhow::Result;
use breez_sdk_liquid::prelude::LiquidSdk;
use breez_sdk_liquid::InputType;
use std::sync::Arc;

async fn parse_input(sdk: Arc<LiquidSdk>) -> Result<()> {
    // ANCHOR: parse-inputs
    let input = "an input to be parsed...";

    match sdk.parse(input).await? {
        InputType::BitcoinAddress { address } => {
            println!("Input is Bitcoin address {}", address.address);
        }
        InputType::Bolt11 { invoice } => {
            println!(
                "Input is BOLT11 invoice for {} msats",
                invoice
                    .amount_msat
                    .map_or("unknown".to_string(), |a| a.to_string())
            );
        }
        InputType::Bolt12Offer { offer, bip353_address } => {
            println!(
                "Input is BOLT12 offer for min {:?} msats - BIP353 was used: {}",
                offer.min_amount,
                bip353_address.is_some()
            );
        }
        InputType::LnUrlPay { data, bip353_address } => {
            println!(
                "Input is LNURL-Pay/Lightning address accepting min/max {}/{} msats - BIP353 was used: {}",
                data.min_sendable, data.max_sendable, bip353_address.is_some()
            );
        }
        InputType::LnUrlWithdraw { data } => {
            println!(
                "Input is LNURL-Withdraw for min/max {}/{} msats",
                data.min_withdrawable, data.max_withdrawable
            );
        }
        // Other input types are available
        _ => {}
    }
    // ANCHOR_END: parse-inputs
    Ok(())
}
